diff -Naur tslib-0.0.2/plugins/Makefile.am tslib-0.0.2.patched/plugins/Makefile.am
--- tslib-0.0.2/plugins/Makefile.am	2006-08-24 23:02:54.000000000 +0200
+++ tslib-0.0.2.patched/plugins/Makefile.am	2009-10-20 13:53:36.000000000 +0200
@@ -107,15 +107,19 @@
   
 variance_la_SOURCES	= variance.c
 variance_la_LDFLAGS	= -module $(LTVSN)
+variance_la_LIBADD	= $(top_builddir)/src/libts.la
   
 dejitter_la_SOURCES	= dejitter.c
 dejitter_la_LDFLAGS	= -module $(LTVSN)
+dejitter_la_LIBADD	= $(top_builddir)/src/libts.la
   
 linear_la_SOURCES	= linear.c
 linear_la_LDFLAGS	= -module $(LTVSN)
+linear_la_LIBADD	= $(top_builddir)/src/libts.la
 
 pthres_la_SOURCES	= pthres.c
 pthres_la_LDFLAGS	= -module $(LTVSN)
+pthres_la_LIBADD	= $(top_builddir)/src/libts.la
 
 # hw access
 corgi_la_SOURCES	= corgi-raw.c
